### PR TITLE
feat(packager): add Node.js bindings, ref #47

### LINKED ISSUE #47: [RAD-53] Node.js bindings
Offering a NPM package for this project would be huge for JavaScript developers that do not want to install Rust and manage dependencies with Cargo. For instance Electron users using `electron-builder` or `electron-packager` could simply install our NPM package and change their build scripts to use it. We could even accept the same configuration objects to simplify the migration process.

This can be achieved with [napi-rs](https://github.com/napi-rs/napi-rs).

<sub>[RAD-54](https://linear.app/crabnebula/issue/RAD-53/[packager]-npm-cli-and-library)<sub>

### PR DESCRIPTION
**Pull request by lucasfernog-crabnebula**
This PR adds a Node.js binding to the packager using [napi.rs](https://napi.rs/). The NPM package is called `@crabnebula/packager`. It includes a CLI and a programatic interface to the `package` function, though the latter was not polished yet and we're focusing on the CLI for now.

The Node.js binding pulls the configuration from `package.json > packager` and comes with Electron support via a plugin that checks if Electron is a dependency, downloads the prebuilt Electron binary and adjust the configuration to work with it. It's been tested on macOS, Linux and Windows.

It does not close #47 yet because we might want to add compatibility layers with electron-build, electron-packager etc so it's easier for users.

What I didn't test yet is the CI workflow that comes with napi-rs for publishing it. That always take some time to figure out. It lives under `.github/workflows/publish-packager-nodejs.yml`.

We might want to come up with a better name for this project, as `cargo-packager` isn't a good name for a project that are also available for NPM in my opinion.

### CODE REVIEW

**Comment by lucasfernog-crabnebula** [comment]
I also added unit tests (also on CI) and it works!

File: crates/packager/src/lib.rs [hunk 1]
```diff
@@ -101,6 +101,35 @@ pub use sign::SigningConfig;
...
+            .max_level_hint()
+            .and_then(|l| l.into_level())
+            .unwrap_or(tracing::Level::INFO),
+        1 => tracing::Level::DEBUG,
+        2.. => tracing::Level::TRACE,
+    }
+}
+
+/// Inits the tracing subscriber.
+pub fn init_tracing_subscriber(verbosity: u8) {
```
Comment (amr-crabnebula) [inline crates/packager/src/lib.rs:117]: I would like to keep the tracing_subscriber out of the library, the node CLI could either depend directly on it or we add this behind a feature flag that is activated on the node CLI

File: crates/packager/src/package/appimage/mod.rs [hunk 2]
```diff
@@ -88,6 +88,8 @@ pub(crate) fn package(ctx: &Context) -> crate::Result<Vec<PathBuf>> {
     // generate deb_folder structure
     tracing::debug!("Generating data");
     let icons = super::deb::generate_data(config, &appimage_deb_data_dir)?;
+    tracing::debug!("Copying files specified in `deb.files`");
+    super::deb::copy_custom_files(config, &appimage_deb_data_dir)?;
```
Comment (amr-crabnebula) [inline crates/packager/src/package/appimage/mod.rs:92]: `deb.files` is something specific to `deb`, don't know if we should use it on `appimage` tbh, why was this needed anyways?

File: crates/packager/src/config/mod.rs [hunk 3]
```diff
@@ -631,8 +631,8 @@ impl Default for LogLevel {
 #[cfg_attr(feature = "schema", derive(schemars::JsonSchema))]
 #[serde(rename_all = "camelCase", deny_unknown_fields)]
 pub struct Binary {
-    /// File name and without `.exe` on Windows
-    pub filename: String,
+    /// Path to the binary. If it's relative, it will be resolved from [`Config::out_dir`].
+    pub path: PathBuf,
```
Comment (amr-crabnebula) [inline crates/packager/src/config/mod.rs:635]: How does this handle `.exe` on Windows? the reason why I didn't use path in the first place is because I wanted users to specify the `filename` and then we add `.exe` on Windows if needed, using a path, breaks this goal.

File: bindings/packager/nodejs/src-ts/plugins/index.ts [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+  if (packageJsonPath === null) {
+    return null;
+  }
+
+  const packageJson = JSON.parse(
+    (await fs.readFile(packageJsonPath)).toString()
+  ) as PackageJson;
+
+  let config = packageJson.packager || null;
```
Comment (amr-crabnebula) [inline bindings/packager/nodejs/src-ts/plugins/index.ts:40]: this could probably be added in the Rust CLI

File: bindings/packager/nodejs/src-ts/plugins/index.ts [hunk 5]
```diff
@@ -0,0 +1,56 @@
+import path from "path";
```
Comment (amr-crabnebula) [inline bindings/packager/nodejs/src-ts/plugins/index.ts:1]: if we are going to add plugins, should we add one for tauri as well? and probably write all in Rust?

File: crates/packager/src/package/appimage/mod.rs (hunk 2 above)
Comment (lucasfernog-crabnebula) [reply crates/packager/src/package/appimage/mod.rs:92]: Electron needs to inject a bunch of files in the appimage. Appimage and deb share the same data folder so I think it's ok, but we can add a separate config for it too.

File: crates/packager/src/config/mod.rs (hunk 3 above)
Comment (lucasfernog-crabnebula) [reply crates/packager/src/config/mod.rs:635]: it's still the same thing, you still just skip the extension, it doesn't change that behavior. but now you can specify absolute paths which is needed for electron

File: bindings/packager/nodejs/src-ts/plugins/index.ts (hunk 4 above)
Comment (lucasfernog-crabnebula) [reply bindings/packager/nodejs/src-ts/plugins/index.ts:40]: package.json reading is specific to Node.js so I think it makes more sense in the binding package instead

File: bindings/packager/nodejs/src-ts/plugins/index.ts (hunk 5 above)
Comment (lucasfernog-crabnebula) [reply bindings/packager/nodejs/src-ts/plugins/index.ts:1]: if I wrote all this in Rust you would tell me not to bloat the cargo packager with NPM stuff 😂 
idk about Tauri, i think most users will stick to tauri-bundler since they are compatible. the problem is just the renaming of the app right? (plus it's harder to detect a tauri project instead of an electron app since Node.js helps us there)

File: crates/packager/src/package/appimage/mod.rs (hunk 2 above)
Comment (amr-crabnebula) [reply crates/packager/src/package/appimage/mod.rs:92]: Yeah let's add a separate config then, better be explicit about it

File: crates/packager/src/config/mod.rs (hunk 3 above)
Comment (amr-crabnebula) [reply crates/packager/src/config/mod.rs:635]: I see, it is my fault, I missed a couple of places where I needed to call `.with_extension("exe")`, pushed a fix

File: bindings/packager/nodejs/src-ts/plugins/index.ts (hunk 5 above)
Comment (amr-crabnebula) [reply bindings/packager/nodejs/src-ts/plugins/index.ts:1]: It is not really about bloat right now, just trying to understand the big picture here, if we are adding plugins, what plugins in the future we might have? why electron gets a special plugin, why dioxus or wry can't have a plugin for it?

File: .changes/packager-fixes.md [hunk 6]
```diff
@@ -0,0 +1,5 @@
+---
+"cargo-packager": patch
+---
+
+Adjustments for `@crabnebula/packager` NPM and bug fixes.
```
Comment (amr-crabnebula) [inline .changes/packager-fixes.md:5]: We can probably remove this change file if it doesn't have any visible changes

File: bindings/packager/nodejs/src-ts/plugins/electron/index.ts [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+
+  const electronPackageJson = JSON.parse(
+    (
+      await fs.readFile(
+        path.resolve(path.dirname(electronPath), "package.json")
+      )
+    ).toString()
+  );
+
+  const zipPath = await downloadElectron(electronPackageJson.version);
```
Comment (amr-crabnebula) [inline bindings/packager/nodejs/src-ts/plugins/electron/index.ts:33]: can't we reuse the one in `node_modules`?

File: bindings/packager/nodejs/src-ts/plugins/electron/index.ts [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+  const appTempPath = await fs.mkdtemp(
+    path.join(os.tmpdir(), packageJson.name || "app-temp")
+  );
+
+  const pruner = new Pruner(appPath, true);
+
+  const outDir = userConfig.outDir ? path.resolve(userConfig.outDir) : null;
+  const ignoredDirs = outDir && outDir !== process.cwd() ? [outDir] : [];
+
+  const filterFunc = (_name: string): boolean => true;
```
Comment (amr-crabnebula) [inline bindings/packager/nodejs/src-ts/plugins/electron/index.ts:56]: this seems like it doesn't filter anything, why is it needed?

File: crates/packager/src/config/mod.rs [hunk 9]
```diff
@@ -631,8 +631,8 @@ impl Default for LogLevel {
 #[cfg_attr(feature = "schema", derive(schemars::JsonSchema))]
 #[serde(rename_all = "camelCase", deny_unknown_fields)]
 pub struct Binary {
-    /// File name and without `.exe` on Windows
-    pub filename: String,
+    /// Path to the binary. If it's relative, it will be resolved from [`Config::out_dir`].
```
Comment (amr-crabnebula) [inline crates/packager/src/config/mod.rs:634]:
```suggestion
    /// Path to the binary (without `.exe` on Windows). If it's relative, it will be resolved from [`Config::out_dir`].
```

File: crates/packager/src/config/mod.rs [hunk 10]
```diff
@@ -889,7 +893,12 @@ impl Config {
 
     /// Returns the out dir
     pub fn out_dir(&self) -> PathBuf {
-        dunce::canonicalize(&self.out_dir).unwrap_or_else(|_| self.out_dir.clone())
+        if self.out_dir.as_os_str().is_empty() {
+            // TODO: we should probably error out when the out dir isn't set
```
Comment (amr-crabnebula) [inline crates/packager/src/config/mod.rs:897]: using the current directory (which will be the directory containing the configuration) here is good enough IMO,

File: bindings/packager/nodejs/src-ts/plugins/electron/index.ts [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+      await fs.chmod(binaryPath, 0o755);
+
+      // make linuxdeploy happy
+      process.env.LD_LIBRARY_PATH = process.env.LD_LIBRARY_PATH
+        ? `${process.env.LD_LIBRARY_PATH}:${zipDir}`
+        : zipDir;
+      // electron needs everything at the same level :)
+      // resources only contains the default_app.asar so we ignore it
+      debianFiles = electronFiles
+        .filter((f) => !["resources"].includes(f))
```
Comment (amr-crabnebula) [inline bindings/packager/nodejs/src-ts/plugins/electron/index.ts:165]:
```suggestion
        .filter((f) => f !== "resources")
```

File: .github/workflows/publish-packager-nodejs.yml [hunk 12]
```diff
@@ -0,0 +1,403 @@
...
+  contents: write
+  id-token: write
+
+on:
+  workflow_dispatch:
+    inputs:
+      releaseId:
+        description: "ID of the `@crabnebula/packager` release"
+        required: true
+  repository_dispatch:
```
Comment (amr-crabnebula) [inline .github/workflows/publish-packager-nodejs.yml:18]: we need to update covector-publish-or-version.yml as well

File: .github/workflows/publish-packager-nodejs.yml [hunk 13]
```diff
@@ -0,0 +1,403 @@
...
+jobs:
+  build:
+    strategy:
+      fail-fast: false
+      matrix:
+        settings:
+          - host: macos-latest
+            target: x86_64-apple-darwin
+            build: |
+              yarn build
```
Comment (amr-crabnebula) [inline .github/workflows/publish-packager-nodejs.yml:34]: this workflow uses yarn while the project uses pnpm

File: .github/workflows/publish-packager-nodejs.yml [hunk 14]
```diff
@@ -0,0 +1,403 @@
...
+          elif git log -1 --pretty=%B | grep "^[0-9]\+\.[0-9]\+\.[0-9]\+";
+          then
+            echo "//registry.npmjs.org/:_authToken=$NPM_TOKEN" >> ~/.npmrc
+            npm publish --tag next --access public
+          else
+            echo "Not a release, skipping publish"
+          fi
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+          NPM_TOKEN: ${{ secrets.NPM_TOKEN }}
```
Comment (amr-crabnebula) [inline .github/workflows/publish-packager-nodejs.yml:403]:
```suggestion
          NPM_TOKEN: ${{ secrets.NPM_TOKEN }}
          RELEASE_ID: ${{ github.event.client_payload.releaseId || inputs.releaseId }}
```

File: bindings/packager/nodejs/package.json [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+  },
+  "engines": {
+    "node": ">= 10"
+  },
+  "scripts": {
+    "artifacts": "napi artifacts",
+    "build-ts": "rm -rf build/ && node generate-config-type.js && tsc",
+    "build": "napi build --platform --profile release-size-optimized && pnpm build-ts",
+    "build:debug": "napi build --platform && pnpm build-ts",
+    "prepublishOnly": "napi prepublish -t npm",
```
Comment (amr-crabnebula) [inline bindings/packager/nodejs/package.json:47]:
```suggestion
    "prepublishOnly": "napi prepublish -t npm --gh-release-id $RELEASE_ID",
```

**Review by amr-crabnebula — CHANGES_REQUESTED** [review]

File: bindings/packager/nodejs/src-ts/plugins/index.ts (hunk 5 above)
Comment (lucasfernog-crabnebula) [reply bindings/packager/nodejs/src-ts/plugins/index.ts:1]: Electron is completely different because it's precompiled so it goes totally against the concepts of the packager, you need to ship a bunch of resources and files by default. It's not 100% controlled by the developer.

File: bindings/packager/nodejs/src-ts/plugins/electron/index.ts (hunk 7 above)
Comment (lucasfernog-crabnebula) [reply bindings/packager/nodejs/src-ts/plugins/electron/index.ts:33]: nope, this downloads the prebuilt electron zip, not the NPM dependency

File: bindings/packager/nodejs/src-ts/plugins/electron/index.ts (hunk 8 above)
Comment (lucasfernog-crabnebula) [reply bindings/packager/nodejs/src-ts/plugins/electron/index.ts:56]: it can be a user-defined filter in the future, I got most of it from https://github.com/electron/packager/blob/main/src/prune.js

File: bindings/packager/nodejs/src-ts/plugins/electron/index.ts (hunk 7 above)
Comment (amr-crabnebula) [reply bindings/packager/nodejs/src-ts/plugins/electron/index.ts:33]: doesn't the NPM dependency also come with a prebuilt binary? anyways I guess it is fine to download but shouldn't we at least cache it?

File: bindings/packager/nodejs/src-ts/plugins/electron/index.ts (hunk 8 above)
Comment (amr-crabnebula) [reply bindings/packager/nodejs/src-ts/plugins/electron/index.ts:56]: if it serves no purpose now, let's remove it and add it later if needed

File: bindings/packager/nodejs/src-ts/plugins/electron/index.ts (hunk 7 above)
Comment (lucasfernog-crabnebula) [reply bindings/packager/nodejs/src-ts/plugins/electron/index.ts:33]: the @electron/get package caches it by default

Comment (amr-crabnebula) [reply bindings/packager/nodejs/src-ts/plugins/electron/index.ts:33]: ah nice

File: bindings/packager/nodejs/src-ts/plugins/index.ts (hunk 5 above)
Comment (amr-crabnebula) [reply bindings/packager/nodejs/src-ts/plugins/index.ts:1]: > Electron is completely different because it's precompiled so it goes totally against the concepts of the packager

That doesn't really go against it, the packager does expect a prebuilt binary

> you need to ship a bunch of resources and files by default.

From packager POV, that's just some configurations that can be set manually by the developer, we are just making it easier for them which I am not against, just seeing if we would do the same for tauri and dioxus.

Comment (lucasfernog) [reply bindings/packager/nodejs/src-ts/plugins/index.ts:1]: Tauri or Dioxus AFAIK does not need any external files to work, but Electron needs you to include their libraries too. If we didn't have this "plugin", every Electron dev would need to download the prebuilt app on their own, figure out the format and what files to include. It's a nightmare and something we can do on our end.
If there's another framework that need this, I think it's ok to add it in the future too. As long as it's easy to detect.

File: .github/workflows/publish-packager-nodejs.yml (hunk 12 above)
Comment (lucasfernog) [reply .github/workflows/publish-packager-nodejs.yml:18]: I didn't do it yet because we have that issue on Tauri.. we didn't find a way to parse the output yet

File: bindings/packager/nodejs/src-ts/plugins/index.ts (hunk 5 above)
Comment (amr-crabnebula) [reply bindings/packager/nodejs/src-ts/plugins/index.ts:1]: > Tauri or Dioxus AFAIK does not need any external files to work,

They do need specific configurations for gtk libraries and `deb` depdendencies and webview2 installation

> If there's another framework that need this, I think it's ok to add it in the future too. As long as it's easy to detect.

yeah that's what I am trying to figure out so we can decide how to make it future-proof, will the packager include plugins or a preconfigured options for specific frameworks to make better DX?

File: .github/workflows/publish-packager-nodejs.yml (hunk 12 above)
Comment (amr-crabnebula) [reply .github/workflows/publish-packager-nodejs.yml:18]: fck, I will create a repo to test again but anyways let's add those changes as well and fix it later in another PR

Comment (amr-crabnebula) [reply .github/workflows/publish-packager-nodejs.yml:18]: I believe I found a fix for it that shouldn't break again https://github.com/amr-crabnebula/covector-test/actions/runs/6896036131/job/18761297111#step:6:1

Comment (amr-crabnebula) [reply .github/workflows/publish-packager-nodejs.yml:18]: basically use inline-js
```yml
      - uses: actions/github-script@v6
        id: set-result
        with:
          result-encoding: string
          script: |
            const output = `${{ toJSON(steps.covector.outputs) }}`;
            const [_, id] = /"-org-dashed-pkg-releaseId": "([0-9]+)"/g.exec(output);
            return id;

      - name: Get result
        run: echo "${{steps.set-result.outputs.result}}"
```

**Comment by lucasfernog-crabnebula** [comment]
publish CI is now ready: https://github.com/crabnebula-dev/cargo-packager/actions/runs/6897523113/job/18766089074

**Review by amr-crabnebula — APPROVED** [review]
